atl-checker/src/algorithms/certain_zero/search_strategy/linear_optimize.rs

```diff
@@ -0,0 +1,515 @@
+                let mut distances: Vec<f32> = Vec::new();
+                for target in &hyperedge.targets {
+                    if let Some(result) = self.get_distance_in_atlvertex(target) {
+                        distances.push(result)
+                    }
+                }
+
+                // If no targets were able to satisfy formula, or something went wrong, return None
+                return if distances.is_empty() {
+                    None
+                } else {
+                    // Find average distance between targets, and return this
+                    let avg_distance = distances.iter().sum::<f32>() / distances.len() as f32;
+                    Some(avg_distance)
+                };
+            }
+            // Same procedure for negation edges as for hyper, just no for loop for all targets, as we only have one target
+            Edge::NEGATION(edge) => {
+                self.get_distance_in_atlvertex(&edge.target)
+            }
+        }
+    }
+
+    /// helper function to iterate through ATLVertices and find distance
+    fn get_distance_in_atlvertex(&mut self, target: &ATLVertex) -> Option<f32> {
+        // TODO change edge by changing order of targets in the edge, based on distance
+        // Find the linear expression from the targets formula, if any
+        let linear_expressions = self.get_linear_expressions_from_atlvertex(target);
+
+        // Polynomials and such not allowed, returns None in such cases
+        if let Some(expressions) = linear_expressions {
+            // get the State in the target
+            let state = self.game.state_from_index(target.state());
+
+            // Check if result is in cache
+            if let Some(cached_result) = self.cache.get(&(expressions.clone(), target.state())) {
+                return Some(*cached_result as f32);
+            }
+
+            let mut expr_distance: f32 = 0.0;
+            for linear_expression in &expressions {
+                // Distance from the state, to fulfilling the linear expression
+                let distance_to_solve_expression = self.minimum_distance_1d(state.clone(), linear_expression);
+                if let Some(distance) = distance_to_solve_expression {
+                    expr_distance = expr_distance + distance;
+                }
+            }
+            // If we got results, add to vec of distances and cache
+            // TODO expand cache with more results?
+            if 0.0 < expr_distance {
+                self.cache.insert((expressions.clone(), target.state()), expr_distance as i32);
+                return Some(expr_distance / expressions.len() as f32);
+            }
+        }
+        None
+    }
+
+    fn get_linear_expressions_from_atlvertex(&self, vertex: &ATLVertex) -> Option<Vec<LinearExpression>> {
+        // get propositions from the formula in the vertex
+        let propositions = vertex.formula().get_propositions_recursively();
```

Comment:
This does not seem correct. This way you do not consider the relation between the propositions/labels. E.g. you don't know if it is `A & B` or `A | B` or `!A & B`.